File: cookbook/persisting-sessions.py
"""
persisting-sessions.py — CopilotForge Cookbook Recipe

Adapted from: https://github.com/github/awesome-copilot/blob/main/cookbook/copilot-sdk/nodejs/persisting-sessions.md

WHAT THIS DOES:
    Shows how to persist Copilot sessions across restarts using custom IDs.
    Create, resume, list, delete sessions and retrieve conversation history.

WHEN TO USE THIS:
    When you need conversations that survive process restarts — chatbots with
    memory, long-running workflows, or multi-step user interactions.

HOW TO RUN:
    1. pip install copilot-sdk
    2. Set your API key:
         bash/zsh:     export COPILOT_API_KEY="your-key-here"
         PowerShell:   $env:COPILOT_API_KEY="your-key-here"
         Windows cmd:  set COPILOT_API_KEY=your-key-here
    3. python cookbook/persisting-sessions.py

PREREQUISITES:
    - Python 3.10+
    - A valid Copilot API key

EXPECTED OUTPUT:
    [Client] Starting...
    [Session] Created: user-123-conversation
    [Assistant] [Echo] You said: "Help me build a REST API in Python." ...
    [Session] Resumed: user-123-conversation (2 messages in history)
    [Assistant] [Echo] You said: "What was I working on?" ...
    [Sessions] 1 active session(s):
      - user-123-conversation (4 messages)
    [History] 4 message(s) in user-123-conversation
    [Session] Deleted: user-123-conversation
    [Client] Stopped.

PLATFORM NOTES:
    - Windows: Use backslashes in paths or os.path.join()
    - macOS/Linux: Forward slashes work natively
    - Environment variables: Use $env:VAR (PowerShell) or export VAR (bash)
"""

from __future__ import annotations

import os
from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional


# --- Types ---


@dataclass
class SessionMessage:
    """A single message in a conversation."""

    role: str  # "user" or "assistant"
    content: str
    timestamp: datetime = field(default_factory=datetime.now)


# --- Mock SDK (replace with real import) ---
# TODO: Replace with: from copilot_sdk import CopilotClient


class CopilotSession:
    """Represents an active Copilot session with message history."""

    def __init__(self, session_id: str, history: list[SessionMessage]) -> None:
        self.id = session_id
        self._history = history

    def send(self, message: str) -> str:
        """Send a message and get a response."""
        self._history.append(SessionMessage(role="user", content=message))
        reply = f'[Echo] You said: "{message}" (session: {self.id})'
        self._history.append(SessionMessage(role="assistant", content=reply))
        return reply

    def get_messages(self) -> list[SessionMessage]:
        """Return a copy of all messages in this session."""
        return list(self._history)

    def destroy(self) -> None:
        """Close the session (does not delete persisted data)."""
        print(f"[Session] Closed: {self.id}")


class CopilotClient:
    """Copilot client with session persistence support."""

    def __init__(self, api_key: str) -> None:
        self._api_key = api_key
        self._store: dict[str, list[SessionMessage]] = {}

    def start(self) -> None:
        print("[Client] Starting...")

    def create_session(self, session_id: str) -> CopilotSession:
        """Create a new session with a custom ID."""
        if session_id in self._store:
            raise RuntimeError(f'Session "{session_id}" already exists. Resume it instead.')
        history: list[SessionMessage] = []
        self._store[session_id] = history
        print(f"[Session] Created: {session_id}")
        return CopilotSession(session_id, history)

    def resume_session(self, session_id: str) -> Optional[CopilotSession]:
        """Resume an existing session by ID. Returns None if not found."""
        history = self._store.get(session_id)
        if history is None:
            return None
        print(f"[Session] Resumed: {session_id} ({len(history)} messages in history)")
        return CopilotSession(session_id, history)

    def list_sessions(self) -> list[dict[str, object]]:
        """List all persisted sessions with their message counts."""
        return [
            {"id": sid, "message_count": len(msgs)}
            for sid, msgs in self._store.items()
        ]

    def delete_session(self, session_id: str) -> bool:
        """Delete a session permanently. Returns True if it existed."""
        if session_id in self._store:
            del self._store[session_id]
            print(f"[Session] Deleted: {session_id}")
            return True
        return False

    def stop(self) -> None:
        print("[Client] Stopped.")


# --- Best practice: meaningful session IDs ---


def build_session_id(user_id: str, context: str) -> str:
    """Create a human-readable, deterministic session ID."""
    return f"{user_id}-{context}"


# --- Main ---


def main() -> None:
    api_key = os.environ.get("COPILOT_API_KEY")
    if not api_key:
        raise RuntimeError("Missing COPILOT_API_KEY environment variable.")

    client = CopilotClient(api_key=api_key)
    client.start()

    try:
        # 1. Create a session with a meaningful custom ID.
        session_id = build_session_id("user-123", "conversation")
        session = client.create_session(session_id)

        response1 = session.send("Help me build a REST API in Python.")
        print("[Assistant]", response1)
        session.destroy()

        # 2. Resume the session later (e.g., after a restart).
        resumed = client.resume_session(session_id)
        if not resumed:
            raise RuntimeError(f'Session "{session_id}" not found — cannot resume.')

        response2 = resumed.send("What was I working on?")
        print("[Assistant]", response2)
        resumed.destroy()

        # 3. List all available sessions.
        sessions = client.list_sessions()
        print(f"[Sessions] {len(sessions)} active session(s):")
        for s in sessions:
            print(f"  - {s['id']} ({s['message_count']} messages)")

        # 4. Get conversation history.
        history_session = client.resume_session(session_id)
        if history_session:
            messages = history_session.get_messages()
            print(f"[History] {len(messages)} message(s) in {session_id}")
            history_session.destroy()

        # 5. Clean up: delete the session permanently.
        client.delete_session(session_id)

    except Exception as e:
        print(f"[Error] {e}")

    finally:
        client.stop()


if __name__ == "__main__":
    main()

File: cookbook/persisting-sessions.ts
/**
 * persisting-sessions.ts — CopilotForge Cookbook Recipe
 *
 * Adapted from: https://github.com/github/awesome-copilot/blob/main/cookbook/copilot-sdk/nodejs/persisting-sessions.md
 *
 * WHAT THIS DOES:
 *   Shows how to persist Copilot sessions across restarts using custom IDs.
 *   Create, resume, list, delete sessions and retrieve conversation history.
 *
 * WHEN TO USE THIS:
 *   When you need conversations that survive process restarts — chatbots with
 *   memory, long-running workflows, or multi-step user interactions.
 *
 * HOW TO RUN:
 *   1. npm install @github/copilot-sdk
 *   2. Set your API key:
 *        bash/zsh:     export COPILOT_API_KEY="your-key-here"
 *        PowerShell:   $env:COPILOT_API_KEY="your-key-here"
 *        Windows cmd:  set COPILOT_API_KEY=your-key-here
 *   3. npx ts-node cookbook/persisting-sessions.ts
 *
 * PREREQUISITES:
 *   - Node.js 18+
 *   - TypeScript 5+
 *   - A valid Copilot API key
 *
 * EXPECTED OUTPUT:
 *   [Client] Starting...
 *   [Session] Created: user-123-conversation
 *   [Assistant] I'll help you build a REST API. Let's start with...
 *   [Session] Resumed: user-123-conversation (2 messages in history)
 *   [Assistant] Continuing where we left off — you wanted a REST API...
 *   [Sessions] 1 active session(s):
 *     - user-123-conversation (2 messages)
 *   [History] 2 message(s) in user-123-conversation
 *   [Session] Deleted: user-123-conversation
 *   [Client] Stopped.
 *
 * PLATFORM NOTES:
 *   - Windows: Use backslashes in paths or path.join()
 *   - macOS/Linux: Forward slashes work natively
 *   - Environment variables: Use $env:VAR (PowerShell) or export VAR (bash)
 */

// --- Types (simulating @github/copilot-sdk) ---

interface CopilotClientConfig {
  apiKey: string;
}

interface SessionMessage {
  role: "user" | "assistant";
  content: string;
  timestamp: Date;
}

interface CopilotSession {
  id: string;
  send(message: string): Promise<string>;
  getMessages(): SessionMessage[];
  destroy(): void;
}

interface CreateSessionOptions {
  sessionId: string;
}

interface CopilotClient {
  start(): void;
  createSession(options: CreateSessionOptions): CopilotSession;
  resumeSession(sessionId: string): CopilotSession | null;
  listSessions(): Array<{ id: string; messageCount: number }>;
  deleteSession(sessionId: string): boolean;
  stop(): void;
}

// --- Mock SDK (replace with real import) ---
// TODO: Replace with: import { CopilotClient } from "@github/copilot-sdk";

function createCopilotClient(config: CopilotClientConfig): CopilotClient {
  const store = new Map<string, SessionMessage[]>();

  function makeSession(sessionId: string, history: SessionMessage[]): CopilotSession {
    return {
      id: sessionId,
      async send(message: string): Promise<string> {
        history.push({ role: "user", content: message, timestamp: new Date() });
        const reply = `[Echo] You said: "${message}" (session: ${sessionId})`;
        history.push({ role: "assistant", content: reply, timestamp: new Date() });
        return reply;
      },
      getMessages(): SessionMessage[] {
        return [...history];
      },
      destroy() {
        console.log(`[Session] Closed: ${sessionId}`);
      },
    };
  }

  return {
    start() {
      console.log("[Client] Starting...");
    },

    createSession(options: CreateSessionOptions): CopilotSession {
      const { sessionId } = options;
      if (store.has(sessionId)) {
        throw new Error(`Session "${sessionId}" already exists. Resume it instead.`);
      }
      const history: SessionMessage[] = [];
      store.set(sessionId, history);
      console.log(`[Session] Created: ${sessionId}`);
      return makeSession(sessionId, history);
    },

    resumeSession(sessionId: string): CopilotSession | null {
      const history = store.get(sessionId);
      if (!history) {
        return null;
      }
      console.log(`[Session] Resumed: ${sessionId} (${history.length} messages in history)`);
      return makeSession(sessionId, history);
    },

    listSessions(): Array<{ id: string; messageCount: number }> {
      return Array.from(store.entries()).map(([id, msgs]) => ({
        id,
        messageCount: msgs.length,
      }));
    },

    deleteSession(sessionId: string): boolean {
      const existed = store.delete(sessionId);
      if (existed) {
        console.log(`[Session] Deleted: ${sessionId}`);
      }
      return existed;
    },

    stop() {
      console.log("[Client] Stopped.");
    },
  };
}

// --- Best practice: meaningful session IDs ---

function buildSessionId(userId: string, context: string): string {
  return `${userId}-${context}`;
}

// --- Main ---

async function main(): Promise<void> {
  const apiKey = process.env.COPILOT_API_KEY;
  if (!apiKey) {
    throw new Error("Missing COPILOT_API_KEY environment variable.");
  }

  const client = createCopilotClient({ apiKey });
  client.start();

  try {
    // 1. Create a session with a meaningful custom ID.
    const sessionId = buildSessionId("user-123", "conversation");
    const session = client.createSession({ sessionId });

    const response1 = await session.send("Help me build a REST API in TypeScript.");
    console.log("[Assistant]", response1);
    session.destroy();

    // 2. Resume the session later (e.g., after a restart).
    const resumed = client.resumeSession(sessionId);
    if (!resumed) {
      throw new Error(`Session "${sessionId}" not found — cannot resume.`);
    }

    const response2 = await resumed.send("What was I working on?");
    console.log("[Assistant]", response2);
    resumed.destroy();

    // 3. List all available sessions.
    const sessions = client.listSessions();
    console.log(`[Sessions] ${sessions.length} active session(s):`);
    for (const s of sessions) {
      console.log(`  - ${s.id} (${s.messageCount} messages)`);
    }

    // 4. Get conversation history.
    const history = client.resumeSession(sessionId);
    if (history) {
      const messages = history.getMessages();
      console.log(`[History] ${messages.length} message(s) in ${sessionId}`);
      history.destroy();
    }

    // 5. Clean up: delete the session permanently.
    client.deleteSession(sessionId);
  } catch (error) {
    console.error("[Error]", error instanceof Error ? error.message : error);
  } finally {
    client.stop();
  }
}

// Run the example.
main().catch(console.error);
